### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

In [1]:
# Dependencies and Setup
import pandas as pd

# File to Load (Remember to Change These)
file_to_load = "Resources/purchase_data.csv"

# Read Purchasing File and store into Pandas data frame
purchase_data = pd.read_csv(file_to_load)

## Player Count

* Display the total number of players


In [2]:
player = purchase_data.loc[:,["Gender", "SN", "Age"]]
player =  player.drop_duplicates()
num_players = player.count()[0]
pd.DataFrame({"Total Players": [num_players]})


Total Players
0            576

## Purchasing Analysis (Total)

* Run basic calculations to obtain number of unique items, average price, etc.


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame


In [3]:
# Run basic calculations
average_item_price = purchase_data["Price"].mean()
total_purchase_value = purchase_data["Price"].sum()
purchase_count = purchase_data["Price"].count()
item_count = len(purchase_data["Item ID"].unique())
# Create a DataFrame to hold results
summary_table = pd.DataFrame({"Number of Unique Items": item_count,
                              "Total Revenue": [total_purchase_value],
                              "Number of Purchases": [purchase_count],
                              "Average Price": [average_item_price]})
# Minor Data Munging
summary_table = summary_table.round(2)
summary_table ["Average Price"] = summary_table["Average Price"].map("${:,.2f}".format)
summary_table ["Number of Purchases"] = summary_table["Number of Purchases"].map("{:,}".format)
summary_table ["Total Revenue"] = summary_table["Total Revenue"].map("${:,.2f}".format)
summary_table = summary_table.loc[:,["Number of Unique Items", "Average Price", "Number of Purchases", "Total Revenue"]]
# Display the summary_table
summary_table



Number of Unique Items Average Price Number of Purchases Total Revenue
0                     179         $3.05                 780     $2,379.77

## Gender Demographics

* Percentage and Count of Male Players


* Percentage and Count of Female Players


* Percentage and Count of Other / Non-Disclosed




In [4]:
total_count = player["Gender"].value_counts()
percentage = total_count / num_players
gender_demographics = pd.DataFrame({"total count": total_count, "percentage of players":
                                   percentage})
gender_demographics['percentage of players'] = gender_demographics['percentage of players'].map("{:,.2%}".format)
gender_demographics


total count percentage of players
Male                           484                84.03%
Female                          81                14.06%
Other / Non-Disclosed           11                 1.91%


## Purchasing Analysis (Gender)

* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. by gender




* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [5]:
purchase_total = purchase_data.groupby(["Gender"]).sum()["Price"].rename("Total Purchase Value")
average = purchase_data.groupby(["Gender"]).mean()["Price"].rename("Average Purchase Price")
counts = purchase_data.groupby(["Gender"]).count()["Price"].rename("Purchase Count")
normalised_total = purchase_total / gender_demographics["total count"]

gender_data = pd.DataFrame({"Purchase Count": counts, "Average Purchase Price": average, "Total Purchase Value":
                            purchase_total, "normalised total": normalised_total})

gender_data["Average Purchase Price"] = gender_data["Average Purchase Price"].map("${:,.2f}".format)
gender_data["Total Purchase Value"] = gender_data["Total Purchase Value"].map("${:,.2f}".format)
gender_data["Purchase Count"] = gender_data["Purchase Count"].map("{:,}".format)
gender_data["Average Total Purchase Per Person"] = gender_data["normalised total"].map("${:,.2f}".format)
gender_data = gender_data.loc[:, ["Purchase Count", "Average Purchase Price", "Total Purchase Value",
                                  "Average Total Purchase Per Person"]]
gender_data


Purchase Count Average Purchase Price  \
Gender                                                        
Female                           113                  $3.20   
Male                             652                  $3.02   
Other / Non-Disclosed             15                  $3.35   

                      Total Purchase Value Average Total Purchase Per Person  
Gender                                                                        
Female                             $361.94                             $4.47  
Male                             $1,967.64                             $4.07  
Other / Non-Disclosed               $50.19                             $4.56

## Age Demographics

* Establish bins for ages


* Categorize the existing players using the age bins. Hint: use pd.cut()


* Calculate the numbers and percentages by age group


* Create a summary data frame to hold the results


* Optional: round the percentage column to two decimal points


* Display Age Demographics Table


In [6]:
#Bins = [<10, 10-14, 15-19, 20-24, 25-29, 30-34, 35-39, 40+]
age_bins = [0, 9.90, 14.90,19.90, 24.90, 29.90, 34.90, 39.90, 99999]
group_names = ["<10", "10-14", "15-19", "20-24", "25-29", "30-34", "35-39", "40+"]
player["Age Ranges"] = pd.cut(player["Age"], age_bins, labels = group_names)
age_totals = player["Age Ranges"].value_counts()
age_percentage = age_totals / num_players
age = pd.DataFrame({"Total Count": age_totals, "Percentage of Players": age_percentage})
age['Percentage of Players'] = age['Percentage of Players'].map("{:,.2%}".format)
age = age.sort_index()
age

Total Count Percentage of Players
<10             17                 2.95%
10-14           22                 3.82%
15-19          107                18.58%
20-24          258                44.79%
25-29           77                13.37%
30-34           52                 9.03%
35-39           31                 5.38%
40+             12                 2.08%

## Purchasing Analysis (Age)

* Bin the purchase_data data frame by age


* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [7]:
purchase_data["Age Ranges"] = pd.cut(purchase_data["Age"], age_bins, labels=group_names)
age_purchase_total = purchase_data.groupby(["Age Ranges"]).sum()["Price"].rename("Total Purchase Value")
age_average = purchase_data.groupby(["Age Ranges"]).mean()["Price"].rename("Average Purchase Price")
age_counts = purchase_data.groupby(["Age Ranges"]).count()["Price"].rename("Purchase Count")
normalized_total = age_purchase_total / age["Total Count"]
age_data = pd.DataFrame({"Purchase Count": age_counts, "Average Purchase Price": age_average, "Total Purchase Value": age_purchase_total, "Normalized Totals": normalized_total})
age_data["Average Purchase Price"] = age_data["Average Purchase Price"].map("${:,.2f}".format)
age_data["Total Purchase Value"] = age_data["Total Purchase Value"].map("${:,.2f}".format)
age_data["Purchase Count"] = age_data["Purchase Count"].map("{:,}".format)
age_data["Avg Total Purchase per Person"] = age_data["Normalized Totals"].map("${:,.2f}".format)
age_data = age_data.loc[:, ["Purchase Count", "Average Purchase Price", "Total Purchase Value", "Avg Total Purchase per Person"]]
age_data

Purchase Count Average Purchase Price Total Purchase Value  \
Age Ranges                                                              
<10                    23                  $3.35               $77.13   
10-14                  28                  $2.96               $82.78   
15-19                 136                  $3.04              $412.89   
20-24                 365                  $3.05            $1,114.06   
25-29                 101                  $2.90              $293.00   
30-34                  73                  $2.93              $214.00   
35-39                  41                  $3.60              $147.67   
40+                    13                  $2.94               $38.24   

           Avg Total Purchase per Person  
Age Ranges                                
<10                                $4.54  
10-14                              $3.76  
15-19                              $3.86  
20-24                              $4.32  
25-29                              $3.81  
30-34                              $4.12  
35-39                              $4.76  
40+                                $3.19

## Top Spenders

* Run basic calculations to obtain the results in the table below


* Create a summary data frame to hold the results


* Sort the total purchase value column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [8]:
purchase_data.groupby("SN")
counts = purchase_data.groupby(["SN"]).count()["Price"].rename("Purchase Count")
purchase_total = purchase_data.groupby(["SN"]).sum()["Price"].rename("Total Purchase Value")
average = purchase_data.groupby(["SN"]).mean()["Price"].rename("Average Purchase Price")
SN = pd.DataFrame({"Purchase Count": counts, "Average Purchase Price": average, 
                   "Total Purchase Value": purchase_total})
SN = SN.sort_values(by=['Total Purchase Value'],ascending=False)

SN["Total Purchase Value"] = SN["Total Purchase Value"].map("${:,.2f}".format)
SN["Purchase Count"] = SN["Purchase Count"].map("{:,}".format)
SN["Average Purchase Price"] = SN["Average Purchase Price"].map("${:,.2f}".format)
SN = SN.loc[:, ["Purchase Count", "Average Purchase Price", "Total Purchase Value"]]
SN.head()


Purchase Count Average Purchase Price Total Purchase Value
SN                                                                    
Lisosia93                5                  $3.79               $18.96
Idastidru52              4                  $3.86               $15.45
Chamjask73               3                  $4.61               $13.83
Iral74                   4                  $3.40               $13.62
Iskadarya95              3                  $4.37               $13.10

## Most Popular Items

* Retrieve the Item ID, Item Name, and Item Price columns


* Group by Item ID and Item Name. Perform calculations to obtain purchase count, item price, and total purchase value


* Create a summary data frame to hold the results


* Sort the purchase count column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [9]:
purchase_data.groupby(["Item ID", "Item Name"])
counts = purchase_data.groupby(["Item ID", "Item Name"]).count()["Price"].rename("Purchase Count")
average = purchase_data.groupby(["Item ID", "Item Name"]).mean()["Price"].rename("Price")
purchase_total = purchase_data.groupby(["Item ID", "Item Name"]).sum()["Price"].rename("Total Purchase Value")
purchase_most_popular_items = pd.DataFrame({"Purchase Count": counts, "Price": average, 
                   "Total Purchase Value": purchase_total})
purchase_most_popular_items = purchase_most_popular_items.sort_values(by=["Purchase Count"],ascending=False)
purchase_most_popular_items["Total Purchase Value"] = purchase_most_popular_items["Total Purchase Value"].map("${:,.2f}".format)
purchase_most_popular_items["Purchase Count"] = purchase_most_popular_items["Purchase Count"].map("{:,}".format)
purchase_most_popular_items["Price"] = purchase_most_popular_items["Price"].map("${:,.2f}".format)
purchase_most_popular_items = purchase_most_popular_items.loc[:, ["Purchase Count", "Price", "Total Purchase Value"]]
purchase_most_popular_items.head()


Purchase Count  Price  \
Item ID Item Name                                                            
92      Final Critic                                             13  $4.61   
178     Oathbreaker, Last Hope of the Breaking Storm             12  $4.23   
145     Fiery Glass Crusader                                      9  $4.58   
132     Persuasion                                                9  $3.22   
108     Extraction, Quickblade Of Trembling Hands                 9  $3.53   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
145     Fiery Glass Crusader                                       $41.22  
132     Persuasion                                                 $28.99  
108     Extraction, Quickblade Of Trembling Hands                  $31.77

## Most Profitable Items

* Sort the above table by total purchase value in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the data frame



In [10]:
purchase_data.groupby(["Item ID", "Item Name"])
counts = purchase_data.groupby(["Item ID", "Item Name"]).count()["Price"].rename("Purchase Count")
average = purchase_data.groupby(["Item ID", "Item Name"]).mean()["Price"].rename("Price")
purchase_total = purchase_data.groupby(["Item ID", "Item Name"]).sum()["Price"].rename("Total Purchase Value")
purchase_most_popular_items = pd.DataFrame({"Purchase Count": counts, "Price": average, 
                   "Total Purchase Value": purchase_total})
purchase_most_popular_items = purchase_most_popular_items.sort_values(by=["Total Purchase Value"],ascending=False)
purchase_most_popular_items["Total Purchase Value"] = purchase_most_popular_items["Total Purchase Value"].map("${:,.2f}".format)
purchase_most_popular_items["Purchase Count"] = purchase_most_popular_items["Purchase Count"].map("{:,}".format)
purchase_most_popular_items["Price"] = purchase_most_popular_items["Price"].map("${:,.2f}".format)
purchase_most_popular_items = purchase_most_popular_items.loc[:, ["Purchase Count", "Price", "Total Purchase Value"]]
purchase_most_popular_items.head()

Purchase Count  Price  \
Item ID Item Name                                                            
92      Final Critic                                             13  $4.61   
178     Oathbreaker, Last Hope of the Breaking Storm             12  $4.23   
82      Nirvana                                                   9  $4.90   
145     Fiery Glass Crusader                                      9  $4.58   
103     Singed Scalpel                                            8  $4.35   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
82      Nirvana                                                    $44.10  
145     Fiery Glass Crusader                                       $41.22  
103     Singed Scalpel                                             $34.80

## Observations & Trends


Firstly, in an analysis such as this I personally like to use a variation in percentage of around 2% to be within the approximated range of a trend.


### Gender Demographics

##### Trend:

Basic Analysis:

- Females make up 14.06% yet make up 15.21% of Total Purchases.
- Males make up 84.03% yet have a lower percentage of Total Purchases at 82.68%.

Although this is disproportionate between male and female players, it is a trend with respect to the Total Purchases. 

Top Spenders repeat the above trend with 4 out of 5 being male.

This is again represented in Most Popular and Most Profitable items, 4 out of the 5 in each are male purchasers. 
 
Given the percentage of players regarding the gender demographics, this is not surprising. However, in a purely financial manner it would make sense to start marketing and making the female population games that would encourage more female participants. 

This is where the trend ends.

###### Break in Trend:

The above trends end when it comes to the Average Purchase Price between male and female players. On average, females ($4.47) and males ($4.07) spend differentiates substantially.

As shown above, females make up a much smaller percentage however their spend is higher than the male demographic and the Average Purchase Price is nearly a whopping 10% higher than their male counterparts. As mentioned above, this is potentially a growth market.


### Age Demographics

##### Trend - Generations Y and Z

For perspective, it is worth to note that the first online game was developed around 1980 where the first 'real' interactive online game (MUD) was developed where another user was able to play the game by connecting through the internet. So when we look at this it is only natural that the trend of online gaming was going to include Generations Y and Z. 

As a result of this we see that 83.20% of Total Purchases were by these Generations. However, it is obvious that those in Gen Y would not start gaming until the evolution of the systems themselves as they grew up.  

This is an expected trend demographically as it is the younger portion of Gen Y (20 - 24) and the older of Gen Z (15 - 19) whom have embraced the evolution of the gaming platform as they themselves have evolved.
